File: persistence-test/src/test/java/org/mariella/test/model/Resource.java
package org.mariella.test.model;

import org.mariella.persistence.runtime.TrackedList;

import javax.persistence.*;
import java.sql.Timestamp;

@javax.persistence.Entity
@Table(name = "RESOURCE")
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
@DiscriminatorColumn(
        name = "TYPE",
        discriminatorType = DiscriminatorType.STRING
)
public class Resource extends Entity {
    private String name;
    private Timestamp lastModified;
    private Folder parent;
    private TrackedList<Resource> children = new TrackedList<>(propertyChangeSupport, "children");

    @Column(name = "NAME")
    public String getName() {
        return name;
    }

    public void setName(String name) {
        propertyChangeSupport.firePropertyChange("name", this.name, this.name = name);
    }

    @ManyToOne
    @JoinColumn(name = "PARENT_ID", referencedColumnName = "ID")
    public Folder getParent() {
        return parent;
    }

    public void setParent(Folder parent) {
        propertyChangeSupport.firePropertyChange("parent", this.parent, this.parent = parent);
    }

    @OneToMany(mappedBy = "parent")
    public TrackedList<Resource> getChildren() {
        return children;
    }

    @Column(name = "LAST_MODIFIED")
    public Timestamp getLastModified() {
        return lastModified;
    }

    public void setLastModified(Timestamp lastModified) {
        propertyChangeSupport.firePropertyChange("lastModified", this.lastModified, this.lastModified = lastModified);
    }

}

File: persistence-kotlin/src/test/kotlin/org/mariella/persistence/kotlin/entities/Resource.kt
package org.mariella.persistence.kotlin.entities

import jakarta.persistence.*
import org.mariella.persistence.runtime.TrackedList
import java.time.Instant

@jakarta.persistence.Entity
@Table(name = "resource_node")
@Inheritance(strategy = InheritanceType.JOINED)
@DiscriminatorColumn(name = "node_type", discriminatorType = DiscriminatorType.STRING)
abstract class Resource : Entity() {
    @get:ManyToOne
    @get:JoinColumn(name = "space_id", referencedColumnName = "id")
    var space: Space? by changeSupport()

    @get:ManyToOne
    @get:JoinColumn(name = "owned_by", referencedColumnName = "id")
    var owner: UserEntity? by changeSupport()

    @get:ManyToOne
    @get:JoinColumn(name = "revision_id", referencedColumnName = "id")
    var revision: Revision? by changeSupport()

    @get:Column(name = "node_comment")
    var comment: String? by changeSupport()

    @get:Column(name = "revision_time")
    var createdAt: Instant by changeSupport()

    @get:Column(name = "locked_at")
    var lockedAt: Instant? by changeSupport()

    @get:Column(name = "entity_id")
    var entityId: String by changeSupport()

    @get:OneToMany(mappedBy = "resource")
    var resourceVersions: MutableList<ResourceVersion> = TrackedList(propertyChangeSupport, "resourceVersions")

    abstract fun isContainer(): Boolean
}